packages/hadron-app-registry/src/register-plugin.tsx

```diff
@@ -160,6 +160,58 @@ const useMockOption = <T extends keyof MockOptions>(key: T): MockOptions[T] => {
   return useContext(MockOptionsContext)[key];
 };
 
+let serviceLocationInProgress = false;
+
+const kLocator = Symbol('CompassServiceLocatorFunction');
+
+/**
+ * Helper method to "create" a service locator function for a particular
+ * service. This method guarantees that service locators are used only in
+ * appropriate context. Every service locator function in the application should
+ * be created with this method.
+ */
+export function createServiceLocator<
+  T extends (this: any, ...args: any[]) => any
+>(fn: T, name = fn.name): T {
+  return Object.assign(
+    function (this: any, ...args: any[]) {
+      if (!serviceLocationInProgress) {
+        throw new Error(
+          `Using service locator function "${name}" outside of the service location lifecycle. ` +
+            `Make sure that service locator function is passed as a second argument to the registerHadronPlugin method and is not used directly in a React render method.`
+        );
+      }
+      return fn.call(this, ...args);
+    },
+    { [kLocator]: true }
+  ) as unknown as T;
+}
+
+/**
+ * Optional service provider creation method. In some cases service providers
+ * need access to other service locators to facilitate service injections. In
+ * these cases service provider can be wrapped with the createServiceProvider
+ * function to allows usage of serviceLocator functions in providers outside of
+ * the usual hadron plugin "activate" lifecycle.
+ */
+export function createServiceProvider<T extends React.FunctionComponent>(
+  fn: T
+): T {
+  const displayName = `ServiceProvider(${fn.displayName ?? fn.name})`;
+  const Provider = function (props: React.ComponentProps<T>) {
+    serviceLocationInProgress = true;
+    const content = fn(props);
+    serviceLocationInProgress = false;
```

Comment:
(same as below re: try/finally)

Reply:
Yes, good point, thanks for catching that!

Reply:
Fixed in 5efff6b6fd892c1827a9c81bd32dbe7ea0b1ed6c